### PR TITLE
support python codes in the imperative model

### PR DESCRIPTION
**Pull request by panyx0718**

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_imperative.py [hunk 1]
```diff
@@ -34,7 +35,35 @@ def forward(self, inputs):
...
 
 
-class MLP(fluid.imperative.PyLayer):
+class MyPyLayer(fluid.imperative.PyLayer):
+    def __init__(self):
+        super(MyPyLayer, self).__init__()
+
+    @staticmethod
+    def forward(inputs):
+        sys.stderr.write('before forward\n')
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_imperative.py:44]: remove the debug info

File: python/paddle/fluid/tests/unittests/test_imperative.py [hunk 2]
```diff
@@ -34,7 +35,35 @@ def forward(self, inputs):
...
+        ret = np.tanh(inputs[0])
+        sys.stderr.write('after forward: %s\n' % ret)
+        tensor = core.LoDTensor()
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_imperative.py:47]: should be fluid.Variable()

File: python/paddle/fluid/tests/unittests/test_imperative.py [hunk 3]
```diff
@@ -34,7 +35,35 @@ def forward(self, inputs):
...
+    def backward(inputs):
+        sys.stderr.write('calling into backward: %s\n' % str(inputs))
+        inp, out, dout = inputs
+        inp = np.array(inp)
+        out = np.array(out)
+        dout = np.array(dout)
+        sys.stderr.write('calling into backward: %s, %s, %s\n' %
+                         (inp, out, dout))
+        ret = np.array(dout) * (1 - np.square(np.array(out)))
+        tensor = core.LoDTensor()
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_imperative.py:61]: same above

File: paddle/fluid/imperative/layer.h [hunk 4]
```diff
@@ -123,16 +132,26 @@ class VarBase {
...
 
   virtual ~OpBase() {
     if (grad_op_desc_) delete grad_op_desc_;
   }
 
   std::map<std::string, std::vector<VarBase*>> ApplyGrad();
 
+  // One of `op_desc_` or `forward_id_` is set, not both.
+  // For pure python PyLayer, use `forward_id_`, otherwise, use op_desc_.
   framework::OpDesc* op_desc_;
```
Comment (velconia) [inline paddle/fluid/imperative/layer.h]: Why don't use PyOpBase here? e.g.. PyOpBase and CppOpBase inherit from OpBase.

File: python/paddle/fluid/imperative/layers.py [hunk 5]
```diff
@@ -37,8 +39,56 @@ def __call__(self, *inputs):
...
+        raise NotImplementedError
+
+    @classmethod
+    def __call__(cls, inputs):
+        tracer = framework._imperative_tracer()
+        block = framework.default_main_program().current_block()
+        inputs = [x._ivar for x in inputs]
+
+        if not hasattr(cls, 'forward_id'):
+            cls.forward_id = core.PyLayer.num_funcs() + 1
```
Comment (velconia) [inline python/paddle/fluid/imperative/layers.py:72]: Consider the multi-thread part, should acquire lock here

File: paddle/fluid/imperative/layer.h [hunk 6]
```diff
@@ -123,16 +132,26 @@ class VarBase {
...
+  int forward_id_;
+  // When has backward, one of `grad_op_desc_` or `backward_id_` is set,
+  // not both.
   framework::OpDesc* grad_op_desc_;
+  int backward_id_;
```
Comment (velconia) [inline paddle/fluid/imperative/layer.h:154]: Why don't use PyOpBase here? e.g.. PyOpBase and CppOpBase inherit from OpBase.

File: paddle/fluid/imperative/layer.h [hunk 7]
```diff
@@ -123,16 +132,26 @@ class VarBase {
...
   virtual ~OpBase() {
     if (grad_op_desc_) delete grad_op_desc_;
   }
 
   std::map<std::string, std::vector<VarBase*>> ApplyGrad();
 
+  // One of `op_desc_` or `forward_id_` is set, not both.
+  // For pure python PyLayer, use `forward_id_`, otherwise, use op_desc_.
   framework::OpDesc* op_desc_;
+  int forward_id_;
```
Comment (velconia) [inline paddle/fluid/imperative/layer.h:150]: Seems ```forward_id``` could be passed directly from the ```py_trace``` func

File: python/paddle/fluid/imperative/layers.py [hunk 8]
```diff
@@ -37,8 +39,56 @@ def __call__(self, *inputs):
...
+            PyLayer.register_func(cls.forward_id, cls.forward)
+            cls.backward_id = core.PyLayer.num_funcs() + 1
+            PyLayer.register_func(cls.backward_id, cls.backward)
+
+        iop = core.OpBase()
+        iop.forward_id = cls.forward_id
+        iop.backward_id = cls.backward_id
+        block.ops.append(iop)
+        ivars = tracer.py_trace(iop, inputs, False)
```
Comment (velconia) [inline python/paddle/fluid/imperative/layers.py:81]: pass forward_id and backward_id directly here

**Review by velconia — APPROVED** [review]
LGTM with little comments

File: python/paddle/fluid/tests/unittests/test_imperative.py (hunk 2 above)
Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_imperative.py:47]: will move this code in parent

File: python/paddle/fluid/tests/unittests/test_imperative.py (hunk 3 above)
Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_imperative.py:61]: will move this code to parent

File: python/paddle/fluid/tests/unittests/test_imperative.py (hunk 1 above)
Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_imperative.py:44]: will do it in next PR

File: paddle/fluid/imperative/layer.h (hunk 6 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.h:154]: I thought about it. But might delay this as code quality clean up.

**Comment by velconia** [comment]
Should add more unit tests for ```PyLayer```,  e.g..

```PyLayer``` and ```Layer``` mixed network

```PyLayer``` MNIST loss v.s.. ```Layer``` MNIST loss

**Pull request merged**